In [94]:
import pandas as pd
import numpy as np
import os

## C1 — Series and DataFrame by hand (2 pts)
Build a 
Series
 of five marks with student names as the index, and print its 
same small table as a 
.index
 and 
.values 
. Build the
DataFrame
 twice — once from a dictionary of lists, once from a list of dictionaries —
and show the two are equal. Print 
.shape 
, 
.columns 
, 
.index 
.

In [2]:
marks_series=pd.Series([85,70,98,60,88],index=['rohan','kartik','reeta','prateek','sujan'])

In [3]:
print(marks_series)
print(marks_series.index)
print(marks_series.values)

rohan      85
kartik     70
reeta      98
prateek    60
sujan      88
dtype: int64
Index(['rohan', 'kartik', 'reeta', 'prateek', 'sujan'], dtype='object')
[85 70 98 60 88]


In [4]:
# DataFrame from a dictionary of lists.
data_dict = {
    "name": ["rohan", "kartik", "reeta","prateek","sujan"],
    "marks": [85,70,98,60,88],
    "city": ["Kathmandu", "Pokhara", "Lalitpur","Damak","Biratnagar"]
}
df_from_dict = pd.DataFrame(data_dict)


In [5]:
# DataFrame from a list of dictionaries.
data_records = [
    {"name": "rohan", "marks": 85, "city": "Kathmandu"},
    {"name": "kartik", "marks": 70, "city": "Pokhara"},
    {"name": "reeta", "marks": 98, "city": "Lalitpur"},
    {"name": "prateek", "marks": 60, "city": "Damak"},
    {"name": "sujan", "marks": 88, "city": "Biratnagar"}
]
df_from_records = pd.DataFrame(data_records)

In [6]:
print("dataframe from a dictionary of lists:")
print(df_from_dict)
print("dataframe frome a list of dictionaries")
print(df_from_records)

dataframe from a dictionary of lists:
      name  marks        city
0    rohan     85   Kathmandu
1   kartik     70     Pokhara
2    reeta     98    Lalitpur
3  prateek     60       Damak
4    sujan     88  Biratnagar
dataframe frome a list of dictionaries
      name  marks        city
0    rohan     85   Kathmandu
1   kartik     70     Pokhara
2    reeta     98    Lalitpur
3  prateek     60       Damak
4    sujan     88  Biratnagar


In [7]:
print("Two dataframes are equal:",df_from_dict.equals(df_from_records))

Two dataframes are equal: True


In [8]:
print("Shape of dataframe:",df_from_dict.shape)
print("Columns:",df_from_dict.columns)
print("Index:",df_from_dict.index)

Shape of dataframe: (5, 3)
Columns: Index(['name', 'marks', 'city'], dtype='object')
Index: RangeIndex(start=0, stop=5, step=1)


## C2 — Load and inspect (3 pts)
Read 
scores_raw.csv 
. Print 
head() 
, 
tail(3) 
, 
shape 
, 
info() 
, 
describe() 
, and
describe(include="object") 
. Then write two to three sentences naming three specific problems visible in
that output, each backed by a number or a value from it

In [9]:
df=pd.read_csv("../data/scores_raw.csv")

In [10]:
df.head()

,student_id,name,city,subject,marks,attendance_percent,exam_date
0,1014,Nisha Gurung,POKHARA,Databases,29.0,81.4,2026-02-22
1,1035,Ojaswi Adhikari,Birgunj,Statistics,52.0,80.4,2026-01-17
2,1036,Prakash Karki,DHARAN,Databases,71.0,89.8,2026-02-07
3,1019,Tara Shrestha,Kathmandu,Machine Learning,76.0,81.4,2026-03-09
4,1001,Aarav Shrestha,kathmandu,Databases,93.0,92.2,2026-01-29


In [11]:
df.tail(3)

,student_id,name,city,subject,marks,attendance_percent,exam_date
123,1016,Prakash Rai,LALITPUR,Statistics,90.0,64.0,2026-01-06
124,1037,Rita Shrestha,kathmandu,Machine Learning,79.0,61.7,2026-01-28
125,1015,Ojaswi Thapa,Biratnagar,Linear Algebra,70.0,70.3,2026-03-20


In [12]:
df.shape

(126, 7)

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126 entries, 0 to 125
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   student_id          126 non-null    int64  
 1   name                126 non-null    object 
 2   city                126 non-null    object 
 3   subject             126 non-null    object 
 4   marks               122 non-null    float64
 5   attendance_percent  116 non-null    float64
 6   exam_date           126 non-null    object 
dtypes: float64(2), int64(1), object(4)
memory usage: 7.0+ KB


In [14]:
df.describe()

,student_id,marks,attendance_percent
count,126.000000,122.000000,116.000000
mean,1020.460317,60.557377,74.413793
std,11.597690,23.807440,12.514033
min,1001.000000,28.000000,52.500000
25%,1010.250000,42.250000,63.125000
50%,1020.500000,54.000000,73.300000
75%,1030.750000,79.750000,84.000000
max,1040.000000,150.000000,98.600000


In [15]:
df.describe(include='object')

,name,city,subject,exam_date
count,126,126,126,126
unique,40,24,5,66
top,Ojaswi Adhikari,Biratnagar,Statistics,2026-01-29
freq,4,10,27,4


In [16]:
#Problems seen in the dataset
# "marks" and "attendance_percent" contain null values
# "marks" contain outliers it have values out of range 1-100
# "exam_date" is stored as object

## C3 — Selecting rows and columns (3 pts)
Print 
df["marks"]
 and 
df[["marks"]]
 with 
type()
 of each, and say in a comment what the difference is.
Then show 
df.loc[0:4, ["name", "subject", "marks"]]
 and 
df.iloc[0:5, 1:4] 
, print how many rows
each returned, and explain in a comment why the two slices treat their end point differently.

In [17]:
df['marks']

0      29.0
1      52.0
2      71.0
3      76.0
4      93.0
       ... 
121    90.0
122    65.0
123    90.0
124    79.0
125    70.0
Name: marks, Length: 126, dtype: float64

In [18]:
type(df['marks'])

pandas.core.series.Series

In [19]:
df[['marks']]

,marks
0,29.0
1,52.0
2,71.0
3,76.0
4,93.0
...,...
121,90.0
122,65.0
123,90.0
124,79.0


In [20]:
type(df[['marks']])

pandas.core.frame.DataFrame

In [21]:
''' 
df['marks'] with single bracket returns series(1-dimensional)
df[['marks']] with double brackets returns dataframe(2 dimensional)
'''

" \ndf['marks'] with single bracket returns series(1-dimensional)\ndf[['marks']] with double brackets returns dataframe(2 dimensional)\n"

In [22]:
loc_slice=df.loc[0:4, ["name", "subject", "marks"]] 
loc_slice

,name,subject,marks
0,Nisha Gurung,Databases,29.0
1,Ojaswi Adhikari,Statistics,52.0
2,Prakash Karki,Databases,71.0
3,Tara Shrestha,Machine Learning,76.0
4,Aarav Shrestha,Databases,93.0


In [23]:
iloc_slice=df.iloc[0:5, 1:4]
iloc_slice

,name,city,subject
0,Nisha Gurung,POKHARA,Databases
1,Ojaswi Adhikari,Birgunj,Statistics
2,Prakash Karki,DHARAN,Databases
3,Tara Shrestha,Kathmandu,Machine Learning
4,Aarav Shrestha,kathmandu,Databases


In [24]:
print("rows returned using loc:",len(loc_slice))
print("rows returned using iloc:",len(iloc_slice))

rows returned using loc: 5
rows returned using iloc: 5


In [25]:
# .loc uses label-based indexing and includes the ending label (4).
# .iloc uses positional indexing and excludes the ending position (5).

## C4 — Boolean filtering (3 pts)
(a) Rows with marks > 80 and subject "Python Programming".

(b) Rows from Kathmandu or Pokhara —note that this under-counts until you clean the city column in C5, and say so. 

 (c) The same filter as (a) writtenwith .query() , proved identical with .equals() .
 
  In a comment, explain why and / or fail inside a bracketfilter but are required inside query() .

In [26]:
# Students who scored above 80 in Python Programming
filter_a=df[(df["marks"] > 80) & (df["subject"] == "Python Programming")]
filter_a

,student_id,name,city,subject,marks,attendance_percent,exam_date
5,1009,Ishan Thapa,Biratnagar,Python Programming,98.0,61.2,2026-02-23
73,1006,Faiz Karki,DHARAN,Python Programming,81.0,98.5,2026-01-05
117,1011,Kiran Adhikari,Birgunj,Python Programming,89.0,78.3,2026-03-24


In [27]:
#students from kathmandu or pokhara
filter_b=df[(df["city"] == "Kathmandu") | (df["city"] == "Pokhara")]
filter_b

,student_id,name,city,subject,marks,attendance_percent,exam_date
61,1014,Nisha Gurung,Pokhara,Python Programming,73.0,NaN,2026-03-06
71,1020,Umesh Gurung,Pokhara,Python Programming,62.0,63.2,2026-01-23
86,1002,Bibek Gurung,Pokhara,Statistics,87.0,67.3,2026-03-21
87,1031,Kiran Shrestha,Kathmandu,Machine Learning,52.0,93.7,2026-01-22
96,1014,Nisha Gurung,Pokhara,Statistics,81.0,96.4,2026-01-08
105,1008,Hari Gurung,Pokhara,Python Programming,44.0,97.1,2026-03-07
114,1001,Aarav Shrestha,Kathmandu,Linear Algebra,93.0,74.4,2026-03-24


In [28]:
filter_a_query=df.query("marks > 80 and subject == 'Python Programming'")
filter_a_query

,student_id,name,city,subject,marks,attendance_percent,exam_date
5,1009,Ishan Thapa,Biratnagar,Python Programming,98.0,61.2,2026-02-23
73,1006,Faiz Karki,DHARAN,Python Programming,81.0,98.5,2026-01-05
117,1011,Kiran Adhikari,Birgunj,Python Programming,89.0,78.3,2026-03-24


In [29]:
print("filter_a and filter_a_query are identical:",filter_a.equals(filter_a_query))

filter_a and filter_a_query are identical: True


In [30]:
# In query(), Python-style 'and' / 'or' are used for conditions.
# In normal DataFrame boolean indexing, '&' / '|' are required because
# Series comparisons are elementwise. Parentheses are also needed around
# individual comparisons when using & or |.|


## C5 — Cleaning inconsistent text (3 pts)
Print df["city"].nunique() and value_counts() before cleaning. Normalise the column with string
methods (strip whitespace, fix capitalisation), then print both again. Checkpoint: 24 distinct spellings
before, 6 after. Re-run your C4(b) filter and report how many rows you were losing

In [31]:
#Before cleaning 
df["city"].nunique()

24

In [32]:
df['city'].value_counts()

city
Biratnagar       10
  Pokhara         9
KATHMANDU         8
LALITPUR          8
kathmandu         7
BIRGUNJ           6
  Kathmandu       6
Birgunj           6
  Dharan          6
pokhara           6
Dharan            6
  Lalitpur        5
Lalitpur          5
Pokhara           5
  Biratnagar      5
dharan            4
lalitpur          4
birgunj           4
biratnagar        4
  Birgunj         3
BIRATNAGAR        3
DHARAN            2
POKHARA           2
Kathmandu         2
Name: count, dtype: int64

In [33]:
#cleaning the inconsistent data
df["city"] = df["city"].astype("string").str.strip().str.title()

In [34]:
#After cleaning
df['city'].nunique()

6

In [35]:
df['city'].value_counts()

city
Kathmandu     23
Pokhara       22
Lalitpur      22
Biratnagar    22
Birgunj       19
Dharan        18
Name: count, dtype: Int64

In [36]:
#After cleaning
#students from kathmandu or pokhara
filter_b_clean=df[(df["city"] == "Kathmandu") | (df["city"] == "Pokhara")]
filter_b_clean

,student_id,name,city,subject,marks,attendance_percent,exam_date
0,1014,Nisha Gurung,Pokhara,Databases,29.0,81.4,2026-02-22
3,1019,Tara Shrestha,Kathmandu,Machine Learning,76.0,81.4,2026-03-09
4,1001,Aarav Shrestha,Kathmandu,Databases,93.0,92.2,2026-01-29
7,1026,Faiz Gurung,Pokhara,Databases,46.0,97.5,2026-03-13
11,1038,Sujan Gurung,Pokhara,Linear Algebra,37.0,70.3,2026-03-09
12,1031,Kiran Shrestha,Kathmandu,Databases,44.0,86.1,2026-02-07
23,1020,Umesh Gurung,Pokhara,Linear Algebra,42.0,73.0,2026-03-26
25,1026,Faiz Gurung,Pokhara,Linear Algebra,28.0,80.4,2026-01-08
32,1013,Manish Shrestha,Kathmandu,Python Programming,67.0,97.3,2026-01-07
35,1019,Tara Shrestha,Kathmandu,Statistics,85.0,67.1,2026-01-09


In [37]:
print("No of rows gained after cleaning:",len(filter_b_clean)-len(filter_b))

No of rows gained after cleaning: 38


## C6 — Missing values (3 pts)
Print 
isna().sum()
 per column. Fill missing 
with missing 
attendance_percent
 with the column mean; drop the rows
marks 
. Print the row count before and after each step. Then, in two sentences, argue whether
mean-filling attendance was a defensible choice here.

In [38]:
print("Missing values per column:")
print(df.isna().sum())

Missing values per column:
student_id             0
name                   0
city                   0
subject                0
marks                  4
attendance_percent    10
exam_date              0
dtype: int64


In [39]:
# Fill missing attendance_percent with its column mean.
rows_before_attendance = len(df)
attendance_mean = df["attendance_percent"].mean()
df["attendance_percent"] = df["attendance_percent"].fillna(attendance_mean)

In [40]:
print("\nAttendance mean used:", attendance_mean)
print("Rows before filling attendance:", rows_before_attendance)
print("Rows after filling attendance:", len(df))


Attendance mean used: 74.41379310344827
Rows before filling attendance: 126
Rows after filling attendance: 126


In [41]:
# Drop rows where marks is missing.
rows_before_marks_drop = len(df)
df1 = df.dropna(subset=["marks"]).copy()


In [42]:
print("\nRows before dropping missing marks:", rows_before_marks_drop)
print("Rows after dropping missing marks:", len(df1))



Rows before dropping missing marks: 126
Rows after dropping missing marks: 122


In [43]:

# Mean-filling attendance is defensible here if the missingness is not strongly related to attendance itself and the missing count is relatively small.
# However, mean filling reduces natural variation and can hide systematic missingness, so it should be reported rather than treated as perfect data

## C7 — Duplicates (3 pts)
Using 
duplicated() 
, print how many fully duplicated rows exist and show a few. Drop them keeping the
first occurrence, and print the shape before and after. Checkpoint: 6 duplicates, 126 → 120 rows.
Separately, report duplicates on the subset 
["student_id", "subject"] 
, and explain in a comment why
that subset is the one that actually matters here

In [44]:
# Fully duplicated rows.
full_dup_mask = df.duplicated(keep=False)
full_dup_count = df.duplicated().sum()


In [45]:
print("Number of fully duplicated rows beyond the first occurrence:", full_dup_count)
print("Duplicated rows including all occurrences:")
display(df.loc[full_dup_mask].head(10))


Number of fully duplicated rows beyond the first occurrence: 6
Duplicated rows including all occurrences:


,student_id,name,city,subject,marks,attendance_percent,exam_date
12,1031,Kiran Shrestha,Kathmandu,Databases,44.0,86.100000,2026-02-07
22,1035,Ojaswi Adhikari,Birgunj,Python Programming,80.0,73.300000,2026-02-21
37,1035,Ojaswi Adhikari,Birgunj,Python Programming,80.0,73.300000,2026-02-21
74,1002,Bibek Gurung,Pokhara,Python Programming,48.0,81.300000,2026-03-13
75,1015,Ojaswi Thapa,Biratnagar,Python Programming,39.0,82.100000,2026-02-06
90,1031,Kiran Shrestha,Kathmandu,Databases,44.0,86.100000,2026-02-07
101,1025,Esha Shrestha,Kathmandu,Statistics,94.0,74.413793,2026-03-17
108,1015,Ojaswi Thapa,Biratnagar,Python Programming,39.0,82.100000,2026-02-06
109,1002,Bibek Gurung,Pokhara,Python Programming,48.0,81.300000,2026-03-13
110,1010,Jyoti Rai,Lalitpur,Statistics,90.0,65.800000,2026-01-27


In [46]:
shape_before_dups = df.shape

In [47]:
df2 = df.drop_duplicates(keep="first")


In [48]:
shape_after_dups = df2.shape

In [49]:
print("\nShape before dropping full duplicates:", shape_before_dups)
print("Shape after dropping full duplicates:", shape_after_dups)


Shape before dropping full duplicates: (126, 7)
Shape after dropping full duplicates: (120, 7)


In [50]:
# Duplicate student_id + subject combinations.
subset_dup_mask = df.duplicated(subset=["student_id", "subject"], keep=False)
subset_dup_count = df.duplicated(subset=["student_id", "subject"]).sum()


In [51]:
print("\nDuplicates based on ['student_id', 'subject']:")
display(df.loc[subset_dup_mask].sort_values(["student_id", "subject"]))


Duplicates based on ['student_id', 'subject']:


,student_id,name,city,subject,marks,attendance_percent,exam_date
74,1002,Bibek Gurung,Pokhara,Python Programming,48.0,81.300000,2026-03-13
109,1002,Bibek Gurung,Pokhara,Python Programming,48.0,81.300000,2026-03-13
110,1010,Jyoti Rai,Lalitpur,Statistics,90.0,65.800000,2026-01-27
121,1010,Jyoti Rai,Lalitpur,Statistics,90.0,65.800000,2026-01-27
75,1015,Ojaswi Thapa,Biratnagar,Python Programming,39.0,82.100000,2026-02-06
108,1015,Ojaswi Thapa,Biratnagar,Python Programming,39.0,82.100000,2026-02-06
101,1025,Esha Shrestha,Kathmandu,Statistics,94.0,74.413793,2026-03-17
113,1025,Esha Shrestha,Kathmandu,Statistics,94.0,74.413793,2026-03-17
12,1031,Kiran Shrestha,Kathmandu,Databases,44.0,86.100000,2026-02-07
90,1031,Kiran Shrestha,Kathmandu,Databases,44.0,86.100000,2026-02-07


In [52]:
print("Duplicate occurrences on student_id + subject:", subset_dup_count)

Duplicate occurrences on student_id + subject: 6


In [53]:
# This subset matters because a student should normally have one score record
# per subject. Exact duplicate rows are only one kind of duplication; two rows
# with the same student_id and subject can still represent conflicting records.

## C8 — Dates (3 pts)
Print 
exam_date 
's dtype, convert it with 
and 
pd.to_datetime() 
, and print the dtype again. Add 
exam_weekday
 columns using the 
exam_month
.dt
 accessor, then filter to exams held after 
2026-02-15
 and print
how many there are.

In [55]:
print("exam_date dtype before conversion:", df["exam_date"].dtype)

exam_date dtype before conversion: object


In [56]:
df["exam_date"] = pd.to_datetime(df["exam_date"], errors="coerce")

In [57]:
print("exam_date dtype after conversion:", df["exam_date"].dtype)

exam_date dtype after conversion: datetime64[ns]


In [58]:
# Extract useful date components with the .dt accessor.
df["exam_weekday"] = df["exam_date"].dt.day_name()
df["exam_month"] = df["exam_date"].dt.month


In [61]:
print("Date columns:")
df[["exam_date", "exam_weekday", "exam_month"]].head()

Date columns:


,exam_date,exam_weekday,exam_month
0,2026-02-22,Sunday,2
1,2026-01-17,Saturday,1
2,2026-02-07,Saturday,2
3,2026-03-09,Monday,3
4,2026-01-29,Thursday,1


In [62]:
after_cutoff = df[df["exam_date"] > pd.Timestamp("2026-02-15")]

In [63]:
print("\nNumber of exams after 2026-02-15:", len(after_cutoff))


Number of exams after 2026-02-15: 64


## C9 — Sorting and renaming (3 pts)
Sort by 
subject
 (A–Z) and, within each subject, 
attendance_percent
 to 
marks
 (high to low); show the top rows. Rename
attendance
 without modifying the original DataFrame, and prove the original
is unchanged by printing both 
.columns 
. Then restore the original row order with 
sort_index() 
.


In [64]:
sorted_df = df.sort_values(
    by=["subject", "attendance_percent", "marks"],
    ascending=[True, False, False]
)


In [65]:
print("Sorted by subject A–Z, then attendance_percent high→low, then marks high→low:")
display(sorted_df.head(10))


Sorted by subject A–Z, then attendance_percent high→low, then marks high→low:


,student_id,name,city,subject,marks,attendance_percent,exam_date,exam_weekday,exam_month
7,1026,Faiz Gurung,Pokhara,Databases,46.0,97.5,2026-03-13,Friday,3
4,1001,Aarav Shrestha,Kathmandu,Databases,93.0,92.2,2026-01-29,Thursday,1
21,1015,Ojaswi Thapa,Biratnagar,Databases,87.0,91.4,2026-01-15,Thursday,1
122,1025,Esha Shrestha,Kathmandu,Databases,65.0,91.1,2026-03-17,Tuesday,3
56,1018,Sujan Karki,Dharan,Databases,63.0,90.6,2026-01-19,Monday,1
2,1036,Prakash Karki,Dharan,Databases,71.0,89.8,2026-02-07,Saturday,2
100,1022,Bibek Rai,Lalitpur,Databases,92.0,89.4,2026-02-17,Tuesday,2
107,1005,Esha Adhikari,Birgunj,Databases,53.0,88.6,2026-02-08,Sunday,2
12,1031,Kiran Shrestha,Kathmandu,Databases,44.0,86.1,2026-02-07,Saturday,2
90,1031,Kiran Shrestha,Kathmandu,Databases,44.0,86.1,2026-02-07,Saturday,2


In [66]:
# Rename without modifying the original DataFrame.
renamed_df = df.rename(columns={"attendance_percent": "attendance"})

In [68]:
print("Renamed DataFrame columns:")
renamed_df.columns

Renamed DataFrame columns:


Index(['student_id', 'name', 'city', 'subject', 'marks', 'attendance',
       'exam_date', 'exam_weekday', 'exam_month'],
      dtype='object')

In [69]:
print("Original DataFrame columns:")
df.columns

Original DataFrame columns:


Index(['student_id', 'name', 'city', 'subject', 'marks', 'attendance_percent',
       'exam_date', 'exam_weekday', 'exam_month'],
      dtype='object')

In [70]:
# Restore original row order/index.
df = df.sort_index()
print("Original row order restored with sort_index().")


Original row order restored with sort_index().


## C10 — Derived columns (3 pts)
Add 
passed
 (
marks >= 40 
), and 
below 40 → 
grade
 using nested 
np.where 
: 85+ → 
A 
, 70–84 → 
B 
, 40–69 → 
,
C
F 
. Print the overall pass rate as a percentage by taking the mean of 
passed
 cast to 
int 
, and
print 
grade.value_counts() 
. Deal with the impossible marks values you found in C2, and say in a comment
what you did with them and why33

In [72]:
# Marks outside 0–100 are physically impossible.
# Convert those values to NaN, then remove those invalid records before
# calculating pass rate and grades. This prevents impossible scores from distorting the derived results.

In [73]:
invalid_mark_mask = ~df["marks"].between(0, 100)
invalid_mark_count = invalid_mark_mask.sum()

In [74]:
print("Impossible marks found:", invalid_mark_count)
print("Impossible mark values:", df.loc[invalid_mark_mask, "marks"].tolist())

Impossible marks found: 6
Impossible mark values: [nan, 150.0, 150.0, nan, nan, nan]


In [75]:
df.loc[invalid_mark_mask, "marks"] = np.nan
df = df.dropna(subset=["marks"]).copy()

In [76]:
# Pass/fail.
df["passed"] = df["marks"] >= 40

In [77]:
# Nested np.where for the required grade bands:
# 85+ = A, 70–84 = B, 40–69 = C, below 40 = F.
df["grade"] = np.where(
    df["marks"] >= 85,
    "A",
    np.where(
        df["marks"] >= 70,
        "B",
        np.where(df["marks"] >= 40, "C", "F")
    )
)

In [78]:
pass_rate = df["passed"].astype(int).mean() * 100

In [79]:
print("\nOverall pass rate: {:.2f}%".format(pass_rate))
print("\nGrade counts:")
print(df["grade"].value_counts())


Overall pass rate: 79.17%

Grade counts:
grade
C    55
F    25
A    23
B    17
Name: count, dtype: int64


In [80]:
df.head(10)

,student_id,name,city,subject,marks,attendance_percent,exam_date,exam_weekday,exam_month,passed,grade
0,1014,Nisha Gurung,Pokhara,Databases,29.0,81.4,2026-02-22,Sunday,2,False,F
1,1035,Ojaswi Adhikari,Birgunj,Statistics,52.0,80.4,2026-01-17,Saturday,1,True,C
2,1036,Prakash Karki,Dharan,Databases,71.0,89.8,2026-02-07,Saturday,2,True,B
3,1019,Tara Shrestha,Kathmandu,Machine Learning,76.0,81.4,2026-03-09,Monday,3,True,B
4,1001,Aarav Shrestha,Kathmandu,Databases,93.0,92.2,2026-01-29,Thursday,1,True,A
5,1009,Ishan Thapa,Biratnagar,Python Programming,98.0,61.2,2026-02-23,Monday,2,True,A
6,1017,Rita Adhikari,Birgunj,Machine Learning,60.0,61.3,2026-03-01,Sunday,3,True,C
7,1026,Faiz Gurung,Pokhara,Databases,46.0,97.5,2026-03-13,Friday,3,True,C
8,1024,Deepika Karki,Dharan,Databases,36.0,72.2,2026-01-09,Friday,1,False,F
10,1003,Chandni Thapa,Biratnagar,Linear Algebra,65.0,75.5,2026-03-30,Monday,3,True,C


## C11 — Types and memory (3 pts)
Convert 
subject
 and 
city
 to the 
category
 dtype. Print 
memory_usage(deep=True)
 for those two
columns before and after, in bytes, and the percentage saved. In a comment, explain what makes 
category
cheaper, and when it would not help

In [81]:
# Measure memory before converting to category.
memory_before = df[["subject", "city"]].memory_usage(deep=True)

In [82]:
print("Memory before category conversion (bytes):")
print(memory_before)

Memory before category conversion (bytes):
Index       960
subject    7478
city       6831
dtype: int64


In [83]:
# Convert repeated text columns to category.
df["subject"] = df["subject"].astype("category")
df["city"] = df["city"].astype("category")

In [84]:
memory_after = df[["subject", "city"]].memory_usage(deep=True)

In [85]:
print("\nMemory after category conversion (bytes):")
print(memory_after)


Memory after category conversion (bytes):
Index      960
subject    604
city       633
dtype: int64


In [86]:
before_total = memory_before.sum()
after_total = memory_after.sum()
saved_bytes = before_total - after_total
saved_percent = (saved_bytes / before_total) * 100

In [88]:
print("\nTotal before:", before_total, "bytes")
print("Total after:", after_total, "bytes")
print("Bytes saved:", saved_bytes)
print("Percentage saved: {:.2f}%".format(saved_percent))


Total before: 15269 bytes
Total after: 2197 bytes
Bytes saved: 13072
Percentage saved: 85.61%


In [89]:
# category is cheaper when a column contains many repeated values because
# Pandas stores the unique values once and stores compact category codes.
# It may not help (and can sometimes use more memory) when most values are unique or the column is very small.

## C12 — Aggregation and a summary table (3 pts)
On the cleaned data, produce: (a) mean, max and count of 
mean 
marks
 by 
city
 and 
subject 
; (c) a 
marks
 per subject using 
groupby().agg() 
; (b)
pivot_table
 of mean marks with cities as rows and subjects as
columns; (d) mean marks per 
exam_month 
, sorted chronologically. Export (a) to 
outputs/summary.csv 
, and
state in one sentence what the month trend shows

In [90]:
# (a) Mean, max and count of marks by city and subject.
summary_city_subject = (
    df.groupby(["city", "subject"], observed=True)["marks"]
      .agg(["mean", "max", "count"])
      .reset_index()
      .sort_values(["city", "subject"])
)

In [91]:
print("(a) Mean, max and count of marks by city and subject:")
display(summary_city_subject)


(a) Mean, max and count of marks by city and subject:


,city,subject,mean,max,count
0,Biratnagar,Databases,80.666667,96.0,3
1,Biratnagar,Linear Algebra,61.666667,70.0,3
2,Biratnagar,Machine Learning,65.600000,89.0,5
3,Biratnagar,Python Programming,50.285714,98.0,7
4,Biratnagar,Statistics,67.666667,93.0,3
5,Birgunj,Databases,53.000000,53.0,1
6,Birgunj,Linear Algebra,59.800000,84.0,5
7,Birgunj,Machine Learning,60.333333,91.0,3
8,Birgunj,Python Programming,73.000000,89.0,4
9,Birgunj,Statistics,60.000000,93.0,5


In [97]:
OUTPUT_DIR = "outputs"
os.makedirs(OUTPUT_DIR, exist_ok=True)

In [98]:
# Export (a) to outputs/summary.csv.
summary_path = os.path.join(OUTPUT_DIR, "summary.csv")
summary_city_subject.to_csv(summary_path, index=False)

print("\nSaved:", summary_path)


Saved: outputs\summary.csv


In [99]:
# (b) Mean marks per subject using groupby().agg().
summary_subject = (
    df.groupby("subject", observed=True)["marks"]
      .agg(mean_marks="mean", max_marks="max", count="count")
      .reset_index()
)

In [100]:
print(" Mean/max/count marks per subject:")
display(summary_subject)

 Mean/max/count marks per subject:


,subject,mean_marks,max_marks,count
0,Databases,56.458333,96.0,24
1,Linear Algebra,51.166667,93.0,24
2,Machine Learning,59.809524,91.0,21
3,Python Programming,58.120000,98.0,25
4,Statistics,69.076923,95.0,26


In [101]:
# (c) Pivot table of mean marks: cities as rows, subjects as columns.
pivot_mean = pd.pivot_table(
    df,
    values="marks",
    index="city",
    columns="subject",
    aggfunc="mean",
    observed=True
)

In [103]:
print("(c) Pivot table — mean marks:")
display(pivot_mean)

(c) Pivot table — mean marks:


subject,Databases,Linear Algebra,Machine Learning,Python Programming,Statistics
city,,,,,
Biratnagar,80.666667,61.666667,65.600000,50.285714,67.666667
Birgunj,53.000000,59.800000,60.333333,73.000000,60.000000
Dharan,63.200000,47.666667,71.333333,78.000000,61.600000
Kathmandu,55.285714,62.000000,61.750000,67.000000,69.666667
Lalitpur,59.000000,46.333333,49.666667,44.250000,89.000000
Pokhara,36.000000,34.250000,45.666667,55.000000,70.333333


In [104]:
# (d) Mean marks per exam_month, sorted chronologically.
month_summary = (
    df.groupby("exam_month", observed=True)["marks"]
      .mean()
      .sort_index()
)

In [105]:
print("(d) Mean marks per exam month:")
display(month_summary)

(d) Mean marks per exam month:


exam_month
1    60.568182
2    57.000000
3    59.166667
Name: marks, dtype: float64

In [106]:
if len(month_summary) >= 2:
    trend_direction = "increases" if month_summary.iloc[-1] > month_summary.iloc[0] else "decreases" if month_summary.iloc[-1] < month_summary.iloc[0] else "stays roughly the same"
    print(
        f"Month trend: mean marks {trend_direction} from the first exam month "
        f"({month_summary.iloc[0]:.2f}) to the last exam month ({month_summary.iloc[-1]:.2f})."
    )
else:
    print("Month trend: there are not enough distinct exam months to describe a trend.")

Month trend: mean marks decreases from the first exam month (60.57) to the last exam month (59.17).



Final cleaned DataFrame shape: (120, 11)
